File: 2021/05/05.hs
{-# LANGUAGE TupleSections #-}
import Data.List.Split
import Data.List
import Control.Arrow

countElems inp = map (head &&& length) $ group $ sort inp

commaNums :: String -> [Int]
commaNums = map read . splitOn ","

toTuples [] = []
toTuples (x:y:xs) = (x,y):toTuples(xs)

removeDiags [] = []
removeDiags ((x,y):(x2,y2):xs)
    | x == x2 || y == y2  = (x,y):(x2,y2):removeDiags(xs)
    | otherwise = removeDiags(xs)

expand ((x,y):(x2,y2):xs)
    | x == x2 = map (x, ) [min y y2 .. max y y2]
    | y == y2 = map (, y) [min x x2 .. max x x2]
    | otherwise = zip (range x x2) (range y y2)
  where
      range x y
        | x < y = [x..y]
        | otherwise = [x,x-1..y]

main = do
    l <- lines <$> readFile "input.txt"
    let input = map (toTuples . commaNums) l
    let noDia = filter (not . null) $ map removeDiags input
    print $ length $ filter (\(x, y) -> y /= 1) $ countElems $ concat $ map expand noDia
    print $ length $ filter (\(x, y) -> y /= 1) $ countElems $ concat $ map expand input

File: 2025/05/05.py
p1 = 0
p2 = 0
a = []
l, r = open("input.txt").read().split("\n\n")
for l in l.split():
    h, j = l.split("-")
    a.append((int(h), int(j)))
a.sort()
c = -1
for e in a:
    p2 += max(0, e[1] + 1 - max(e[0], c))
    c = max(c, e[1] + 1)

for l in map(int, r.split()):
    for e in a:
        if e[0] <= l and e[1] >= l:
            p1 += 1
            break
print(p1)
print(p2)
